common/wireaddr.c

```diff
@@ -197,4 +323,215 @@ bool parse_wireaddr(const char *arg, struct wireaddr *addr, u16 defport,
+{
+return io_read(conn, reach->buffer,7,io_tor_connect_make_onion, reach);
+}
+
+
+static struct io_plan *io_tor_connect_authenticate(struct io_conn *conn, struct tor_service_reaching *reach)
+{
+		sprintf((char *)reach->buffer,"AUTHENTICATE %s\r\n",(char *)reach->cookie);
+
+		if (reach->noauth)
+				sprintf((char *)reach->buffer,"AUTHENTICATE\r\n");
+
+		reach->hlen = strlen((char *)reach->buffer);
+
+return io_write(conn, reach->buffer,  reach->hlen, io_tor_connect_after_authenticate, reach);
+
 }
+
+
+
+static struct io_plan *io_tor_connect_after_answer_pi(struct io_conn *conn, struct tor_service_reaching *reach)
+{
+	char *p,*p2;
+	int i;
+	static u8 buf[MAX_TOR_COOKIE_LEN];
+
+	reach->noauth=false;
+
+	if (strstr((char *)reach->buffer,"NULL")) reach->noauth=true;
+	else
+	if (strstr((char *)reach->buffer,"HASHEDPASSWORD") && (reach->ld->tor_service_password != NULL))
+				{
+					reach->noauth=false;
+					sprintf((char *)reach->cookie,"\"%s\"",reach->ld->tor_service_password);
+				}
+	else
+	if((p = strstr((char *)reach->buffer,"COOKIEFILE=")))
+	{
+	assert(strlen(p) > 12);
+	p2 = strstr((char *)(p+12),"\"");
+	assert(p2 != NULL);
+	i=strlen(p2);
+	*(char *)(p +(strlen(p)-i)) = 0;
+	int fd = open((char *)(p+12), O_RDONLY );
+
+	if (fd < 0)
+	return io_close(conn);
+	if (!read(fd, &buf, sizeof(buf)))
+		 {
+			 close(fd);
+			 return io_close(conn);
+		 } else  close(fd);
+
+	hex_encode(buf, 32, (char *)reach->cookie, 80);
+	reach->noauth=false;
+	}
+	else
+	return io_close(conn);
+
+	return io_tor_connect_authenticate(conn,reach);
+
+}
+
+static struct io_plan *io_tor_connect_after_protocolinfo(struct io_conn *conn, struct tor_service_reaching *reach)
+{
+
+	memset(reach->buffer,0,MAX_TOR_SERVICE_READBUFFER_LEN);
+	return io_read_partial(conn, reach->buffer, MAX_TOR_SERVICE_READBUFFER_LEN ,&reach->hlen, &io_tor_connect_after_answer_pi, reach);
+}
+
+static struct io_plan *io_tor_connect_after_resp_to_connect(struct io_conn *conn, struct tor_service_reaching *reach)
+{
+
+			 sprintf((char *)reach->buffer,"PROTOCOLINFO 1\r\n");
+		  	 reach->hlen = strlen((char *)reach->buffer);
+
+	return io_write(conn, reach->buffer,  reach->hlen, io_tor_connect_after_protocolinfo, reach);
+}
+
+
+static struct io_plan *tor_connect_finish(struct io_conn *conn,
+		struct tor_service_reaching *reach)
+{
+	return io_tor_connect_after_resp_to_connect(conn,reach);
+};
+
+static struct io_plan *tor_conn_init(struct io_conn *conn, struct lightningd *ld)
+{
+	static struct tor_service_reaching reach;
+	static struct addrinfo *ai_tor;
+	reach.ld = ld;
+	getaddrinfo(tal_strdup(NULL,"127.0.0.1"),tal_strdup(NULL,"9051"), NULL,&ai_tor);
+
+	return io_connect(conn, ai_tor, &tor_connect_finish, &reach);
+}
+
+bool create_tor_hidden_service_conn(struct lightningd *ld)
+{
+	int fd;
+	struct io_conn *conn;
+
+	fd = socket(AF_INET, SOCK_STREAM, 0);
+	conn = io_new_conn(NULL, fd, &tor_conn_init,ld);
+	if (!conn)
+		exit(1);
```

Comment:
Use `errx` instead... `errx(1, "cannot create new TOR connection");`.  This is somewhat drastic though and it is better to return a failure --- what is the point of having a `bool` return type here?